File: app/src/main/java/com/boco/whl/funddemo/module/main/fragment/MainFragment.java
package com.boco.whl.funddemo.module.main.fragment;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.util.DisplayMetrics;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.EditText;
import android.widget.GridView;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.ScrollView;

import com.boco.whl.funddemo.R;
import com.boco.whl.funddemo.module.activity.SearchActivity;
import com.boco.whl.funddemo.module.activity.annndroid.LoadImageActivity;
import com.boco.whl.funddemo.module.activity.annndroid.MarqueeActivity;
import com.boco.whl.funddemo.module.activity.annndroid.baidumap.LocationActivity;
import com.boco.whl.funddemo.module.activity.annndroid.eventbus.EventBusTestActivity;
import com.boco.whl.funddemo.module.activity.annndroid.fusionchart.FusionChartTestActivity;
import com.boco.whl.funddemo.module.activity.annndroid.glide.GlideTestActivity;
import com.boco.whl.funddemo.module.activity.annndroid.imagecompress.ImageCompressActivity;
import com.boco.whl.funddemo.module.activity.annndroid.listview.ListTest;
import com.boco.whl.funddemo.module.activity.annndroid.mvp.view.UserActivity;
import com.boco.whl.funddemo.module.activity.annndroid.mvpV2.CustomerActivity;
import com.boco.whl.funddemo.module.activity.annndroid.rxjava.RxGDActivity;
import com.boco.whl.funddemo.module.activity.annndroid.rxjava.RxImageActivity;
import com.boco.whl.funddemo.module.activity.annndroid.rxjava.RxjavaTestActivity;
import com.boco.whl.funddemo.module.adapter.CategoryItemAdapter;
import com.boco.whl.funddemo.utils.IntentUT;
import com.bumptech.glide.Glide;
import com.orhanobut.logger.Logger;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;

/**
 * 首页
 *
 * @author Administrator
 */
public class MainFragment extends Fragment {
    private static final String ARG_PARAM1 = "param1";
    private static final String ARG_PARAM2 = "param2";
    @BindView(R.id.queryimage)
    ImageView queryimage;
    @BindView(R.id.search_text)
    EditText searchText;
    @BindView(R.id.category)
    GridView category;
    Unbinder unbinder;
    @BindView(R.id.guide_iv)
    ImageView guideIv;
    @BindView(R.id.tipRL)
    RelativeLayout tipRL;
    @BindView(R.id.mainScrollView)
    ScrollView mainScrollView;

    private String mParam1;
    private String mParam2;

    private OnFragmentInteractionListener mListener;
    private Intent intent;
    /**
     *
     */
    private float initTouchY = 0;
    private DisplayMetrics metrics;
    private boolean mScaling = false;

    public MainFragment() {
    }

    public static MainFragment newInstance(String param1, String param2) {
        MainFragment fragment = new MainFragment();
        Bundle args = new Bundle();
        args.putString(ARG_PARAM1, param1);
        args.putString(ARG_PARAM2, param2);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mParam1 = getArguments().getString(ARG_PARAM1);
            mParam2 = getArguments().getString(ARG_PARAM2);
        }

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_first, container, false);
        unbinder = ButterKnife.bind(this, view);
        initCategory();
        //
        initScrollScale();

        return view;
    }

    private void initScrollScale() {
        metrics = new DisplayMetrics();
        getActivity().getWindowManager().getDefaultDisplay().getMetrics(metrics);
        mainScrollView.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                ViewGroup.LayoutParams lp = (ViewGroup.LayoutParams) guideIv.getLayoutParams();
                switch (event.getAction()) {
                    case MotionEvent.ACTION_UP:
                        mScaling = false;
                        lp.width = metrics.widthPixels;
                        lp.height = metrics.widthPixels * 237 / 421;
                        guideIv.setLayoutParams(lp);
                        break;
                    case MotionEvent.ACTION_MOVE:
                        int upY = v.getScrollY();
                        float touchY = event.getY();
                        if (!mScaling) {
                            if (upY == 0) {
                                initTouchY = event.getY();
                            } else {
                                break;
                            }
                        }
                        float deltaY = touchY - initTouchY;
                        if (deltaY < 0) {
                            break;
                        }
                        mScaling = true;
                        lp.width = (int) (metrics.widthPixels + deltaY * 0.45);
                        lp.height = (int) ((metrics.widthPixels + deltaY * 0.45) * 237 / 421);
                        guideIv.setLayoutParams(lp);

                        return true;
                    default:
                        break;
                }
                return false;
            }
        });
        mainScrollView.scrollTo(0, 0);

    }

    private void initCategory() {
        String[] titles = {"Glide", "RxJava", "EventBus", "RxImage", "RxOperator", "Retrofit", "OkHttp"
                , "Baidu", "Map", "FusionChart", "MVP1", "MVP2", "kotlin", "listview",
                "imgCompress", "marquee", "loadImage"
                , "updating", "updating", "updating", "updating", "updating", "updating", "updating"
                , "updating", "updating", "updating", "updating", "updating", "updating", "updating"
                , "updating", "updating", "updating", "updating", "updating", "updating", "updating"
                , "updating", "updating", "updating", "updating", "updating", "updating", "updating"
        };
        Logger.d(titles);
        CategoryItemAdapter adapter = new CategoryItemAdapter(getActivity(), titles);
        category.setAdapter(adapter);
        category.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
                switch (i) {
                    case 0:
                        intent = new Intent(getActivity(), GlideTestActivity.class);
                        break;
                    case 1:
                        //rxjava 广东接口
                        intent = new Intent(getActivity(), RxGDActivity.class);
                        break;
                    case 2:
                        intent = new Intent(getActivity(), EventBusTestActivity.class);
                        break;
                    case 3:
                        //rxjava 图片测试
                        intent = new Intent(getActivity(), RxImageActivity.class);
                        break;
                    case 4:
                        intent = new Intent(getActivity(), RxjavaTestActivity.class);
                        break;
                    case 5:
                        intent = new Intent(getActivity(), RxImageActivity.class);
                        break;
                    case 7:
                        intent = new Intent(getActivity(), SearchActivity.class);
                        break;
                    case 8:
                        intent = new Intent(getActivity(), LocationActivity.class);
                        break;
                    case 9:
                        intent = new Intent(getActivity(), FusionChartTestActivity.class);
                        break;
                    case 10:
                        intent = new Intent(getActivity(), UserActivity.class);
                        break;
                    case 11:
                        intent = new Intent(getActivity(), CustomerActivity.class);
                        break;
                    case 12:
                        break;
                    case 13:
                        intent = new Intent(getActivity(), ListTest.class);
                        break;
                    case 14:
                        intent = new Intent(getActivity(), ImageCompressActivity.class);
                        break;
                    case 15:
                        intent = new Intent(getActivity(), MarqueeActivity.class);
                        break;
                    case 16:
                        intent = new Intent(getActivity(), LoadImageActivity.class);
                        break;
                    default:
                        break;
                }
                startActivity(intent);
            }
        });
    }

    public void onButtonPressed(Uri uri) {
        if (mListener != null) {
            mListener.onFragmentInteraction(uri);
        }
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener) {
            mListener = (OnFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @OnClick(R.id.queryimage)
    public void onViewClicked() {
        IntentUT.getInstance().openActivity(getActivity(), SearchActivity.class, false);
    }


    public interface OnFragmentInteractionListener {
        void onFragmentInteraction(Uri uri);
    }
}

File: app/src/main/java/com/boco/whl/funddemo/module/fragment/MainFragment.kt
package com.boco.whl.funddemo.module.fragment

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.DisplayMetrics
import android.view.*
import android.view.ViewTreeObserver.OnGlobalLayoutListener
import android.widget.*
import androidx.fragment.app.Fragment
import com.boco.whl.funddemo.R
import com.boco.whl.funddemo.module.activity.my.SearchActivity
import com.boco.whl.funddemo.module.activity.thirdlib.LoadImageActivity
import com.boco.whl.funddemo.module.activity.thirdlib.MarqueeActivity
import com.boco.whl.funddemo.module.activity.thirdlib.eventbus.EventBusTestActivity
import com.boco.whl.funddemo.module.activity.thirdlib.ffmpeg.HelloJni
import com.boco.whl.funddemo.module.activity.thirdlib.ffmpeg.PlayActivity
import com.boco.whl.funddemo.module.activity.thirdlib.glide.GlideTestActivity
import com.boco.whl.funddemo.module.activity.thirdlib.listview.ListTest
import com.boco.whl.funddemo.module.activity.thirdlib.mvp.view.UserActivity
import com.boco.whl.funddemo.module.activity.thirdlib.mvpV2.CustomerActivity
import com.boco.whl.funddemo.module.activity.thirdlib.retrofit.DownloadApkActivity
import com.boco.whl.funddemo.module.activity.thirdlib.rxjava.RxImageActivity
import com.boco.whl.funddemo.module.activity.thirdlib.rxjava.RxjavaTestActivity
import com.boco.whl.funddemo.module.adapter.CategoryItemAdapter
import kotlinx.android.synthetic.main.fragment_first.*
import java.util.*

/**
 * 首页
 *
 * @author Administrator
 */
class MainFragment : Fragment() {

    private var mParam1: String? = null
    private var mParam2: String? = null
    private var mListener: OnFragmentInteractionListener? = null
    private var intent: Intent? = null

    /**
     *
     */
    private var initTouchY = 0f
    private var metrics: DisplayMetrics? = null
    private var mScaling = false
    private val mMarqueeViews: MutableList<View> = ArrayList()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            mParam1 = arguments!!.getString(ARG_PARAM1)
            mParam2 = arguments!!.getString(ARG_PARAM2)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_first, container, false)
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initCategory()
        initScrollScale()
        initMarquee()
    }

    /**
     * 头条布局
     */
    private fun initMarquee() {
        val data: MutableList<String> = ArrayList()
        data.add("美团卓拙百词斩1614  技责精，github精")
        data.add("基础精,文档精，深步精，源码精，步骤精")
        data.add("实践精,log精，翻墙精，博客精，新敏精")
        mMarqueeViews.clear() //记得加这句话，不然可能会产生重影现象
        var i = 0
        while (i < data.size) {
            val position = i
            //设置滚动的单个布局
            val moreView = LayoutInflater.from(activity).inflate(R.layout.item_view, null) as LinearLayout
            //初始化布局的控件
            val tv1 = moreView.findViewById<View>(R.id.tv1) as TextView
            val tv2 = moreView.findViewById<View>(R.id.tv2) as TextView
            /**
             * 设置监听
             */
            moreView.findViewById<View>(R.id.rl).setOnClickListener { Toast.makeText(activity, position.toString() + "你点击了" + data[position], Toast.LENGTH_SHORT).show() }
            /**
             * 设置监听
             */
            moreView.findViewById<View>(R.id.rl2).setOnClickListener {
                if (data.size > position + 1) {
                    Toast.makeText(activity, position.toString() + "你点击了" + data[position + 1], Toast.LENGTH_SHORT).show()
                } else {
                    Toast.makeText(activity, position.toString() + "你点击了" + data[0], Toast.LENGTH_SHORT).show()
                }
            }
            //进行对控件赋值
            tv1.text = data[i]
            if (data.size > i + 1) {
                //奇数条
                tv2.text = data[i + 1]
            } else {
                //偶数条
                //因为淘宝那儿是两条数据，但是当数据是奇数时就不需要赋值第二个，所以加了一个判断，还应该把第二个布局给隐藏掉
                //moreView.findViewById(R.id.rl2).setVisibility(View.GONE);
                //修改了最后一个没有 将第一个拼接到最后显示
                tv2.text = data[0]
            }

            //添加到循环滚动数组里面去
            mMarqueeViews.add(moreView)
            upView!!.setViews(mMarqueeViews)
            i = i + 2
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun initScrollScale() {
        metrics = DisplayMetrics()
        activity!!.windowManager.defaultDisplay.getMetrics(metrics)
        val lp = guide_iv!!.layoutParams
        lp.width = metrics!!.widthPixels
        lp.height = metrics!!.widthPixels * 237 / 421
        guide_iv!!.layoutParams = lp
        mainScrollView!!.setOnTouchListener { v: View, event: MotionEvent ->
            when (event.action) {
                MotionEvent.ACTION_UP -> {
                    mScaling = false
                    lp.width = metrics!!.widthPixels
                    lp.height = metrics!!.widthPixels * 237 / 421
                    guide_iv!!.layoutParams = lp
                }
                MotionEvent.ACTION_MOVE -> {
                    val upY = v.scrollY
                    val touchY = event.y
                    if (!mScaling) {
                        initTouchY = if (upY == 0) {
                            event.y
                        } else {
                            0f
                        }
                    }
                    val deltaY = touchY - initTouchY
                    if (deltaY < 0) {
                        0f
                    }

                    //缩放代码
                    mScaling = true
                    lp.width = (metrics!!.widthPixels + deltaY * 0.45).toInt()
                    lp.height = ((metrics!!.widthPixels + deltaY * 0.45) * 237 / 421).toInt()
                    guide_iv!!.layoutParams = lp
                    return@setOnTouchListener true
                }
                else -> {
                }
            }
            false
        }
        val observer = guide_iv!!.viewTreeObserver
        observer.addOnGlobalLayoutListener(object : OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                if (guide_iv != null) {
                    guide_iv!!.viewTreeObserver.removeOnGlobalLayoutListener { onGlobalLayout() }
                    val imageHeight = guide_iv!!.height
                    //增加上滑显示操作
                    mainScrollView!!.setScrollViewListener { observableScrollView, x, y, oldX, oldY ->
                        if (y <= 0) {
                            tipLL!!.alpha = 0f
                        } else if (y < imageHeight) {
                            tipLL!!.alpha = y.toFloat() / imageHeight
                        } else {
                            tipLL!!.alpha = 1f
                        }
                    }
                }
            }
        })
        mainScrollView!!.scrollTo(0, 0)
    }

    private fun initCategory() {
        val titles = arrayOf("Glide", "RxJava", "EventBus", "RxImage", "RxOperator", "Retrofit", "OkHttp", "Baidu", "Map", "FusionChart", "MVP1", "MVP2", "kotlin", "listview",
                "imgCompress", "marquee", "loadImage", "JNI", "FFMPEG", "RetrofitDownload", "updating",
                "updating"
        )
        val adapter = CategoryItemAdapter(activity, titles)
        category?.let {
            category!!.adapter = adapter
            category!!.onItemClickListener = AdapterView.OnItemClickListener { adapterView: AdapterView<*>?, view: View?, i: Int, l: Long ->
                when (i) {
                    0 -> intent = Intent(activity, GlideTestActivity::class.java)
                    1 ->                     //rxjava 广东接口
                        intent = Intent(activity, EventBusTestActivity::class.java)
                    2 -> intent = Intent(activity, EventBusTestActivity::class.java)
                    3 ->                     //rxjava 图片测试
                        intent = Intent(activity, RxImageActivity::class.java)
                    4 -> intent = Intent(activity, RxjavaTestActivity::class.java)
                    5 -> intent = Intent(activity, RxjavaTestActivity::class.java)
                    7 -> intent = Intent(activity, SearchActivity::class.java)
                    8 -> intent = Intent(activity, SearchActivity::class.java)
                    9 -> intent = Intent(activity, UserActivity::class.java)
                    10 -> intent = Intent(activity, UserActivity::class.java)
                    11 -> intent = Intent(activity, CustomerActivity::class.java)
                    12 -> {
                    }
                    13 -> intent = Intent(activity, ListTest::class.java)
                    14 -> intent = Intent(activity, PlayActivity::class.java)
                    15 -> intent = Intent(activity, MarqueeActivity::class.java)
                    16 -> intent = Intent(activity, LoadImageActivity::class.java)
                    17 -> intent = Intent(activity, HelloJni::class.java)
                    18 -> intent = Intent(activity, PlayActivity::class.java)
                    19 -> intent = Intent(activity, DownloadApkActivity::class.java)
                    else -> {
                    }
                }
                if (intent != null) {
                    startActivity(intent)
                }
            }
        }
    }

    fun onButtonPressed(uri: Uri?) {
        if (mListener != null) {
            mListener!!.onFragmentInteraction(uri)
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        mListener = if (context is OnFragmentInteractionListener) {
            context
        } else {
            throw RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
    }

    override fun onDestroyView() {
        super.onDestroyView()
    }


    interface OnFragmentInteractionListener {
        fun onFragmentInteraction(uri: Uri?)
    }

    companion object {
        private const val ARG_PARAM1 = "param1"
        private const val ARG_PARAM2 = "param2"
        fun newInstance(param1: String?, param2: String?): MainFragment {
            val fragment = MainFragment()
            val args = Bundle()
            args.putString(ARG_PARAM1, param1)
            args.putString(ARG_PARAM2, param2)
            fragment.arguments = args
            return fragment
        }
    }
}